### PR TITLE
mm/iob: Support alloc IOB via malloc

### PR DESCRIPTION
**Pull request by zhhyu7**
## Summary
Support the network interface card driver to receive zero copies of packets and send and receive giant frame packets, allowing drivers to initialize the DMA buffer to the iob structure, and we can apply for IOB with large memory.
## Testing
sim:local

### CODE REVIEW

File: include/nuttx/mm/iob.h [hunk 1]
```diff
@@ -88,10 +88,18 @@
...
+#  define IOB_BUFSIZE(p) ((p)->io_bufsize)
+#else
+#  define IOB_BUFSIZE(p) CONFIG_IOB_BUFSIZE
+#endif
+
 /****************************************************************************
  * Public Types
  ****************************************************************************/
 
+typedef CODE void (*iob_free_t)(FAR void *data);
```
Comment (anchao) [inline include/nuttx/mm/iob.h:101]:
```suggestion
typedef CODE void (*iob_free_cb_t)(FAR struct iob_s *iob);
```

File: mm/iob/iob_alloc.c [hunk 2]
```diff
@@ -197,6 +200,23 @@ static FAR struct iob_s *iob_allocwait(bool throttled, unsigned int timeout)
...
+ *   Dummy free callback function, do nothing.
+ *
+ * Input Parameters:
+ *   data -
+ *
+ ****************************************************************************/
+
+static void iob_free_dummy(FAR void *data)
+{
+}
```
Comment (anchao) [inline mm/iob/iob_alloc.c:217]:
```suggestion
static void iob_free_dynamic(FAR struct iob_s *iob)
{
  kmm_free(iob);
}
```

File: mm/iob/iob_alloc.c [hunk 3]
```diff
@@ -329,3 +349,89 @@ FAR struct iob_s *iob_tryalloc(bool throttled)
...
+ *
+ *   size    - The size of the data parameter
+ *   free_cb - Notify the caller when the iob is freed. The caller can
+ *             perform additional operations on the data before it is freed.
+ *             The free_cb is called when the iob is freed.
+ *
+ ****************************************************************************/
+
+FAR struct iob_s *iob_alloc_with_data(FAR void *data, uint16_t size,
+                                      iob_free_t free_cb)
```
Comment (anchao) [inline mm/iob/iob_alloc.c:419]: free_cb should not be NULL

File: mm/iob/iob_free.c [hunk 4]
```diff
@@ -112,6 +115,15 @@ FAR struct iob_s *iob_free(FAR struct iob_s *iob)
               next, next->io_pktlen, next->io_len);
     }
 
+#ifdef CONFIG_IOB_ALLOC
+  if (iob->io_free != NULL)
+    {
+      iob->io_free(iob->io_data);
+      kmm_free(iob);
```
Comment (anchao) [inline mm/iob/iob_free.c:122]:
```suggestion
      iob->io_free(iob);
```

Comment (zhhyu7) [reply mm/iob/iob_free.c:122]: @anchao we need to free iob head when driver call iob_alloc_with_data with custom free callback.

File: include/nuttx/mm/iob.h (hunk 1 above)
Comment (zhhyu7) [reply include/nuttx/mm/iob.h]: done

File: mm/iob/iob_alloc.c (hunk 3 above)
Comment (zhhyu7) [reply mm/iob/iob_alloc.c]: done

**Review by anchao — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**